Can't build on Arch linux (FFmpeg)

**Build Log**
What is the build error?
```
make
[ 90%] Built target lpmove
[ 90%] Built target lpoptions
[ 90%] Built target lpstat
[ 90%] Linking C shared library IconServices
ld: warning: static initializer found in 'CMakeFiles/IconServices.dir/src/IconServices.m.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
ld: warning: static initializer found in 'CMakeFiles/IconServices.dir/src/IconServices.m.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
[ 91%] Built target IconServices
make[1]: *** [CMakeFiles/Makefile2:40843: src/CoreAudio/AudioToolbox/CMakeFiles/AudioToolbox.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 91%] Linking C shared library AppKit
ld: warning: instance method 'drawWithRect:options:' in category from CMakeFiles/AppKit.dir/NSStringDrawing.m.o conflicts with same method from another category
ld: warning: instance method 'size' in category from CMakeFiles/AppKit.dir/NSStringDrawing.m.o conflicts with same method from another category
ld: warning: instance method 'drawInRect:' in category from CMakeFiles/AppKit.dir/NSStringDrawing.m.o conflicts with same method from another category
ld: warning: instance method 'drawAtPoint:' in category from CMakeFiles/AppKit.dir/NSStringDrawing.m.o conflicts with same method from another category
[ 92%] Built target AppKit
make: *** [Makefile:136: all] Error 2
```

**System Information**
What system are you building with?

| Software | Version |
| --- | --- |
| Clang | X.Y.Z |
| CMake | X.Y.Z |
| Linux Kernel | X.Y.Z |
| Darling | Git Commit Hash |

If you did a multi-threaded build, you would have to look for it in the output (try searching for any reference to "Error").

```
make[1]: *** [CMakeFiles/Makefile2:40843: src/CoreAudio/AudioToolbox/CMakeFiles/AudioToolbox.dir/all] Error 2
```
This error above only tells me that an error occurred in AudioToolbox, it's not enough information for me to know what is actually causing the error.

If you are not sure where the actual error is located, I strongly recommend post the entire output (either in a pastebin, or equivalent site.

Okay, here is the cause:

```
[ 84%] Building C object src/external/security/SecurityTool/macOS/CMakeFiles/securitytool_macos.dir/keychain_lock.c.o
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:103:8: error: no member named 'channels' in 'AVCodecContext'
  103 |                 cIn->channels = inSourceFormat->mChannelsPerFrame;
      |                 ~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:106:52: error: no member named 'channels' in 'AVCodecContext'
  106 |                 std::cout << "Converting from PCM with " << cIn->channels << " channels at " << cIn->sample_rate << " Hz\n";
      |                                                             ~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:134:13: error: no member named 'channels' in 'AVCodecContext'
  134 |         m_encoder->channels = m_destinationFormat.mChannelsPerFrame;
      |         ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:136:13: error: no member named 'channel_layout' in 'AVCodecContext'
  136 |         m_encoder->channel_layout = CAChannelCountToLayout(m_destinationFormat.mChannelsPerFrame);
      |         ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:164:16: error: no member named 'channel_layout' in 'AVFrame'
  164 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |         ~~~~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:164:44: error: no member named 'channel_layout' in 'AVCodecContext'
  164 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |                                        ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:167:78: error: no member named 'channels' in 'AVCodecContext'
  167 |         int audioSampleBuffer_size = av_samples_get_buffer_size(nullptr, m_encoder->channels, m_audioFrame->nb_samples, m_encoder->sample_fmt, 0);
      |                                                                          ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:177:66: error: no member named 'channels' in 'AVCodecContext'
  177 |         if (int err = avcodec_fill_audio_frame(m_audioFrame, m_encoder->channels, m_encoder->sample_fmt,
      |                                                              ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:456:4: warning: TODO: test this new avcodec decoder code [-W#warnings]
  456 |                 #warning TODO: test this new avcodec decoder code
      |                  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:570:5: warning: TODO: test this new avcodec encoder code [-W#warnings]
  570 |                         #warning TODO: test this new avcodec encoder code
      |                          ^
2 warnings and 8 errors generated.
```

Hmm...

```
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:150:13: error: no member named 'channel_layout' in 'AVCodecContext'
  150 |         m_encoder->channel_layout = CAChannelCountToLayout(m_destinationFormat.mChannelsPerFrame);
      |         ~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:178:16: error: no member named 'channel_layout' in 'AVFrame'
  178 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |         ~~~~~~~~~~~~  ^
/home/jmss/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:178:44: error: no member named 'channel_layout' in 'AVCodecContext'
  178 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |                                        ~~~~~~~~~  ^
```

I also encountered the original issue in a Manjaro / Arch Linux build environment. I tried using bc553383735c1ace4b544751037ecbd724665edc refs/heads/ffmpeg_fix for this build but still encountered an error.

```text
7 warnings and 3 errors generated.
2 warnings generated.
make[2]: *** [src/CoreAudio/AudioToolbox/CMakeFiles/AudioToolbox.dir/build.make:104: src/CoreAudio/AudioToolbox/CMakeFiles/AudioToolbox.dir/AudioConverterImpl.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:40445: src/CoreAudio/AudioToolbox/CMakeFiles/AudioToolbox.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Earlier in the build
```text
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:103:2: warning: "TODO: Remove deprecated 'channels' once we no longer support older distros"  [-W#warnings]
  103 | #warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
      |  ^
[ 84%] Building C object src/external/apr/apr-util/apr-util/CMakeFiles/aprutil.dir/buckets/apr_buckets_refcount.c.o
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:143:2: warning: "TODO: Remove deprecated 'channels' once we no longer support older distros"  [-W#warnings]
  143 | #warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
      |  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:150:13: error: no member named 'channel_layout' in 'AVCodecContext'
  150 |         m_encoder->channel_layout = CAChannelCountToLayout(m_destinationFormat.mChannelsPerFrame);
      |         ~~~~~~~~~  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:178:16: error: no member named 'channel_layout' in 'AVFrame'
  178 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |         ~~~~~~~~~~~~  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:178:44: error: no member named 'channel_layout' in 'AVCodecContext'
  178 |         m_audioFrame->channel_layout = m_encoder->channel_layout;
      |                                        ~~~~~~~~~  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:180:2: warning: "TODO: Remove deprecated 'channels' once we no longer support older distros"  [-W#warnings]
  180 | #warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
      |  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:479:4: warning: TODO: test this new avcodec decoder code [-W#warnings]
  479 |                 #warning TODO: test this new avcodec decoder code
      |                  ^
/home/stuarts/git/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:593:5: warning: TODO: test this new avcodec encoder code [-W#warnings]
  593 |                         #warning TODO: test this new avcodec encoder code
```

| Software | Version |
| --- | --- |
| Clang | 18.1.18 |
| CMake | 3.30.0 |
| Linux Kernel | 6.9.9-1-MANJARO |
| Darling | bc553383735c1ace4b544751037ecbd724665edc refs/heads/ffmpeg_fix |

I could resolve the no member named 'channel_layout' in '...' error by using ch_layout instead but now I get this error:

/home/user/darling/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp:150:23: error: no viable overloaded '='
  150 |         m_encoder->ch_layout = CAChannelCountToLayout(m_destinationFormat.mChannelsPerFrame);

I'm not smart enough to solve the problem but I hope this helps.

@CuriousTommy are you on manjaro unstable, so it's more like Arch?

Still compile error:

```
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regcomp.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regenc.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regerror.c.o
2 warnings generated.
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regexec.c.o
[ 85%] Linking C shared library SpotlightServices
ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
ld: warning: static initializer found in 'CMakeFiles/SpotlightServices.dir/src/SpotlightServices.m.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regparse.c.o
ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
ld: warning: static initializer found in 'CMakeFiles/SpotlightServices.dir/src/SpotlightServices.m.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/regsyntax.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/ruby.c.o
[ 85%] Built target SpotlightServices
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/safe.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/signal.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/sprintf.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/st.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/strftime.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/string.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/struct.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/symbol.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/thread.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/time.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/transient_heap.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/transcode.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/util.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/variable.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/version.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/vm_backtrace.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/vm_dump.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/vm_trace.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/vm.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/missing/explicit_bzero.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/missing/setproctitle.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/enc/ascii.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/enc/unicode.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/enc/us_ascii.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/enc/utf_8.c.o
[ 85%] Building C object src/external/ruby/CMakeFiles/Ruby.dir/ruby/enc/trans/newline.c.o
[ 85%] Linking C shared library Ruby
ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
ld: warning: -undefined dynamic_lookup is incompatible with dyld share cache
ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
ld: warning: -undefined dynamic_lookup is incompatible with dyld share cache
[ 85%] Built target Ruby
make: *** [Makefile:136: all] Error 2
```

> are you on manjaro unstable, so it's more like Arch?

Not sure, just went with the default install.

> Still compile error:
> ```
> ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
> ld: warning: -undefined dynamic_lookup is incompatible with dyld share cache
> ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
> ld: warning: -undefined dynamic_lookup is incompatible with dyld share cache
> [ 85%] Built target Ruby
> make: *** [Makefile:136: all] Error 2
> ```

Again... the log you posted doesn't tell me what is causing the `Error 2` to occur. If you are not sure where the actual error is at, post the entire log in a Pastebin/GitHub gist.

--- a/src/CoreAudio/AFAVFormatComponent/AudioFileFormatGeneric.cpp
+++ b/src/CoreAudio/AFAVFormatComponent/AudioFileFormatGeneric.cpp
@@ -92,15 +92,7 @@ UncertainResult AudioFileFormatGeneric::FileDataIsThisFormat(UInt32 inDataByteSi
 	probeData.buf = buf.data();
 	probeData.buf_size = inDataByteSize;
 
-#warning "TODO: Remove old `read_probe` call once we no longer support older distros" 
-#if LIBAVCODEC_VERSION_MAJOR >= 61
 	return av_probe_input_format(&probeData, false) != nullptr ? kTrue : kFalse;
-#else
-	const AVInputFormat* fmt = av_find_input_format(m_avformatShortName);
-	if (!fmt)
-		return false;
-	return fmt->read_probe(&probeData) ? kTrue : kFalse;
-#endif
 }
 
 AudioFileObject* AudioFileFormatGeneric::New()

--- a/src/CoreAudio/AFAVFormatComponent/AudioFileFormatGeneric.cpp
+++ b/src/CoreAudio/AFAVFormatComponent/AudioFileFormatGeneric.cpp
@@ -81,10 +81,6 @@ void AudioFileFormatGeneric::GetFileTypeName(CFStringRef *outName)
 
 UncertainResult AudioFileFormatGeneric::FileDataIsThisFormat(UInt32 inDataByteSize, const void* inData)
 {
-	const AVInputFormat* fmt = av_find_input_format(m_avformatShortName);
-	if (!fmt)
-		return false;
-	
 	std::vector<uint8_t> buf;
 	AVProbeData probeData;
 
@@ -96,7 +92,15 @@ UncertainResult AudioFileFormatGeneric::FileDataIsThisFormat(UInt32 inDataByteSi
 	probeData.buf = buf.data();
 	probeData.buf_size = inDataByteSize;
 
+#warning "TODO: Remove old `read_probe` call once we no longer support older distros" 
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+	return av_probe_input_format(&probeData, false) != nullptr ? kTrue : kFalse;
+#else
+	const AVInputFormat* fmt = av_find_input_format(m_avformatShortName);
+	if (!fmt)
+		return false;
 	return fmt->read_probe(&probeData) ? kTrue : kFalse;
+#endif
 }
 
 AudioFileObject* AudioFileFormatGeneric::New()

--- a/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp
+++ b/src/CoreAudio/AudioToolbox/AudioConverterImpl.cpp
@@ -100,10 +100,19 @@ OSStatus AudioConverter::create(const AudioStreamBasicDescription* inSourceForma
 	
 	if (inSourceFormat->mFormatID == kAudioFormatLinearPCM)
 	{
+#warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+		cIn->ch_layout.nb_channels = inSourceFormat->mChannelsPerFrame;
+#else
 		cIn->channels = inSourceFormat->mChannelsPerFrame;
+#endif
 		cIn->sample_rate = inSourceFormat->mSampleRate;
-		
+
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+		std::cout << "Converting from PCM with " << cIn->ch_layout.nb_channels << " channels at " << cIn->sample_rate << " Hz\n";
+#else
 		std::cout << "Converting from PCM with " << cIn->channels << " channels at " << cIn->sample_rate << " Hz\n";
+#endif
 	}
 	
 	if (avcodec_open2((*out)->m_decoder, codecIn, nullptr) < 0)
@@ -131,9 +140,15 @@ void AudioConverter::initEncoder()
 	
 	m_encoder->codec_type = AVMEDIA_TYPE_AUDIO;
 	m_encoder->bit_rate = m_outBitRate;
+#warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+	m_encoder->ch_layout.order = AV_CHANNEL_ORDER_UNSPEC;
+	m_encoder->ch_layout.nb_channels = m_destinationFormat.mChannelsPerFrame;
+#else
 	m_encoder->channels = m_destinationFormat.mChannelsPerFrame;
-	m_encoder->sample_rate = m_destinationFormat.mSampleRate;
 	m_encoder->channel_layout = CAChannelCountToLayout(m_destinationFormat.mChannelsPerFrame);
+#endif
+	m_encoder->sample_rate = m_destinationFormat.mSampleRate;
 	m_encoder->sample_fmt = CACodecSampleFormat(&m_destinationFormat);
 
 #ifdef DEBUG_AUDIOCONVERTER
@@ -161,10 +176,16 @@ void AudioConverter::allocateBuffers()
 	
 	m_audioFrame->nb_samples = ENCODER_FRAME_SAMPLES;
 	m_audioFrame->format = m_encoder->sample_fmt;
-	m_audioFrame->channel_layout = m_encoder->channel_layout;
 
-	
+#warning "TODO: Remove deprecated 'channels' once we no longer support older distros" 
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+	m_audioFrame->ch_layout.order = m_encoder->ch_layout.order;
+	m_audioFrame->ch_layout.nb_channels = m_encoder->ch_layout.nb_channels;
+	int audioSampleBuffer_size = av_samples_get_buffer_size(nullptr, m_encoder->ch_layout.nb_channels, m_audioFrame->nb_samples, m_encoder->sample_fmt, 0);
+#else
+	m_audioFrame->channel_layout = m_encoder->channel_layout;
 	int audioSampleBuffer_size = av_samples_get_buffer_size(nullptr, m_encoder->channels, m_audioFrame->nb_samples, m_encoder->sample_fmt, 0);
+#endif
 	void* audioSampleBuffer = (uint8_t*) av_malloc(audioSampleBuffer_size);
 
 	if (!audioSampleBuffer)
@@ -174,8 +195,13 @@ void AudioConverter::allocateBuffers()
 	}
 
 	// Setup the data pointers in the AVFrame
+#if LIBAVCODEC_VERSION_MAJOR >= 61
+	if (int err = avcodec_fill_audio_frame(m_audioFrame, m_encoder->ch_layout.nb_channels, m_encoder->sample_fmt,
+		(const uint8_t*) audioSampleBuffer, audioSampleBuffer_size, 0 ); err < 0)
+#else
 	if (int err = avcodec_fill_audio_frame(m_audioFrame, m_encoder->channels, m_encoder->sample_fmt,
 		(const uint8_t*) audioSampleBuffer, audioSampleBuffer_size, 0 ); err < 0)
+#endif
 	{
 		std::cerr << "AudioConverter::allocateBuffers(): Could not set up audio frame\n";
 		throw std::runtime_error("AudioConverter::allocateBuffers(): Could not set up audio frame");

